src/components/tl/mlx5/tl_mlx5_wqe.c

```diff
@@ -0,0 +1,306 @@
+                                 struct mlx5_wqe_umr_pointer_seg *pseg,
+                                 uint32_t ptr_mkey, void *ptr_address,
+                                 int *xlat_size, uint64_t *reglen)
+{
+    uint64_t                              byte_count = 0;
+    struct mlx5_wqe_umr_repeat_block_seg *rb;
+    struct mlx5_wqe_umr_repeat_ent_seg *  eb;
+    int                                   i;
+
+    /* set pointer segment */
+    pseg->mkey    = htobe32(ptr_mkey);
+    pseg->address = htobe64((uint64_t)ptr_address);
+
+    /* set actual repeated and entry blocks segments */
+    rb               = ptr_address;
+    rb->op           = htobe32(0x400); // PRM header entry - repeated blocks
+    rb->reserved     = 0;
+    rb->num_ent      = htobe16(num_interleaved);
+    rb->repeat_count = htobe32(repeat_count);
+    eb               = rb->entries;
+
+    /*
+     * ------------------------------------------------------------
+     * | repeat_block | entry_block | entry_block |...| entry_block
+     * ------------------------------------------------------------
+     */
+    for (i = 0; i < num_interleaved; i++, eb++) {
+        byte_count += data[i].bytes_count;
+        eb->va         = htobe64(data[i].addr);
+        eb->byte_count = htobe16(data[i].bytes_count);
+        eb->stride     = htobe16(data[i].bytes_count + data[i].bytes_skip);
+        eb->memkey     = htobe32(data[i].lkey);
+    }
+
+    rb->byte_count = htobe32(byte_count);
+    *reglen        = byte_count * repeat_count;
+    *xlat_size     = (num_interleaved + 1) * sizeof(*eb);
+}
+
+ucc_status_t ucc_tl_mlx5_post_umr(struct ibv_qp *     qp,
+                                  struct mlx5dv_mkey *dv_mkey,
+                                  uint32_t access_flags, uint32_t repeat_count,
+                                  uint16_t num_entries,
+                                  struct mlx5dv_mr_interleaved *data,
+                                  uint32_t ptr_mkey, void *ptr_address)
+{
+    uint32_t opcode = MLX5_OPCODE_UMR;
+    uint64_t reglen = 0;
+    uint32_t opmode = 0x0;
+    uint32_t n_ds   = (sizeof(struct mlx5_wqe_ctrl_seg) +
+                       sizeof(struct mlx5_wqe_umr_ctrl_seg) +
+                       sizeof(struct mlx5_wqe_mkey_context_seg) +
+                       sizeof(struct mlx5_wqe_umr_pointer_seg)) /
+                       DS_SIZE;
+    uint8_t fm_ce_se =
+        MLX5_WQE_CTRL_INITIATOR_SMALL_FENCE | MLX5_WQE_CTRL_CQ_UPDATE;
+    struct ibv_qp_ex *                qp_ex = ibv_qp_to_qp_ex(qp);
+    struct mlx5dv_qp_ex *             mqp = mlx5dv_qp_ex_from_ibv_qp_ex(qp_ex);
+    struct mlx5_wqe_ctrl_seg *        ctrl;
+    struct mlx5_wqe_umr_ctrl_seg *    umr_ctrl_seg;
+    struct mlx5_wqe_mkey_context_seg *mk_seg;
+    struct mlx5_wqe_umr_pointer_seg * pseg;
+    char                              wqe_desc[n_ds * DS_SIZE];
+    int                               xlat_size;
+
+    ibv_wr_start(qp_ex);
```

Comment:
do you need to protect all the below code here? or you can move ibv_wr_start() just before WQE builder mlx5dv_wr_raw_wqe at line 195 to reduce the critical section?